### DISCUSSION TITLE
how to distribute with a package?

### DISCUSSION THREAD

**Discussion by greg-minshall** [glouw/ctl#20] [General]
hi.  let's say i wanted to distribute a source package that (very happily!) relies on `ctl`.  how should i package it?

i'm pretty sure there is no, e.g., Arch Linux package for `ctl`.

i could use git (gasp!) `submodule`.

or ... any thoughts?

(a second, somewhat related, thought: maybe `#includes` should be `#include <ctl/vec.h>` rather than straight `#include <vec.h>`?  that would lessen the likelihood of file name collisions, which would probably be good.)

**Comment by glouw** [comment]
Considering CTL is header only there's no reason for it to be a package - I don't want it to become something complicated like C++'s Boost. It's standalone enough to just drop the `ctl` folder in your project and ship:

![image](https://user-images.githubusercontent.com/7000616/149581261-22a3ad76-e851-4f81-a36f-f0411cc59cfb.png)

You can compile with `-Isys ctl` to keep the `<>` brackets when including CTL modules as I believe it may hide a couple annoying warnings that would serve only to clutter your build output. Otherwise, just include directly what you need after you drop the ctl folder into your project:

```
#define P
#define T int
#include "ctl/vec.h"
```

**Reply by greg-minshall** [reply]
Gustav, thanks.  i went ahead and added it as a `git submodule`:

````
git submodule add https://github.com/glouw/ctl.git
````

just in the hopes that that will help upgrading in the future.  (though, sigh, maybe it makes it harder for people cloning my project; something to document in my README.org.)

i'm not sure what `-Isys ctl` would do.  but, i do have `-Ictl -Ictl/ctl`, and then `#include <ctl/vec.h>` from my code.  (the `-Ictl/ctl` is for the `#includes` in the library itself.)

cheers.

**Reply by glouw** [reply]
Ah, I guess it's just `-Ictl` then. It's been a year since I've touched systems related C. You're on the right track. Best of luck with the release

**Reply by greg-minshall** [reply]
thanks.  i put one target in my makefile to detect uninitialized submodules, and another to do the initialization.  here, just for reference:
````
# has our submodule been initialized?  if not, issue an appropriate
# message and terminate the make
checksubmodules:
	@if (git submodule status | grep '^-' > /dev/null); then \
		echo "ctl submodule not initialized; run \"make submodules\""; \
		exit 1; \
	fi

# get submodules (well, just `ctl`) initialized
submodules:
	git submodule init
	git submodule update
````

**Reply by glouw** [reply]
That's really cool. I'm glad github submodules make for a happy alternative to pacman. It's kinda like an impromptu rust package manager at this point. Once you release, do you mind sharing what you've built? I'd love to see how you incorporated CTL

**Reply by greg-minshall** [reply]
here is something: [rx-cli](https://git.sr.ht/~minshall/rx-cli).  it's still pretty preliminary.  and, probably shows un-tutored use of `ctl`! :)  cheers.
